### ISSUE TITLE
[Flutter 3.7.0] Sending Class Objects between ```FlutterEngines``` created from ```FlutterEngineGroup``` produces errors after hot restart.

### ISSUE THREAD

**Issue by dayoul** [flutter/flutter#119589]
Currently I have 2 different ```FlutterEngine``` made from a same ```FlutterEngineGroup```.
My app is using ```SendPort``` to communicate between engines(isolates), and it works fine even if I send any dart instances(```sendPort.send(MyInstance())```) before hot restart. 
But after hot restart, I get error like below and only basic dart type is allowed to be sent(int, String...).
```
Invalid argument: is a regular instance: Instance of 'Data'
```
In this issue(#114005), which I have created, I have found a workaround to use  ```IsolateNameServer``` to store ```SendPort``` of each ```FlutterEngine``` respectively, and it worked fine on Flutter version 3.3.5 event after hot restart.
But it's not working again on Flutter version 3.7.0 after hot restart.
Hot restart function is really critical to my team and is one of the reason why we chose Flutter for our project.

Please check the [repository](https://github.com/dayoul/flutter_engine_group_error) to reproduce this issue.

Result of ```flutter doctor -v```
[✓] Flutter (Channel stable, 3.7.0, on macOS 13.1 22C65 darwin-x64, locale ko-KR) 

## Why do I need multiple flutter engines
Basically the app that my team is building is a messenger app. So in many circumstances, FCM push or APNS push gets received on background and to store that push data, we run ```FlutterEngine``` from background without screen in order to access DB or TCP instance in Flutter(dart) side.

## Why do I have to send Class Instances between ```FlutterEngine```s(```Isolate```s)
Background ```FlutterEngine``` mentioned above is in charge of lots of jobs that are not related to UI(business logic, etc). So it would be really complex if just making it to pass only primitive data types.

**Comment by exaby73** [comment]
Hello @dayoul. Why not serialize your classes to a JSON string and then send it? It would be easy to do with packages like json_serializable generating the `fromJson` and `toJson` methods for you

**Comment by dayoul** [comment]
@exaby73 , serializing classes generates huge amount of boilerplate code(```fromJson```, ```toJson```). It would be inefficient especially for large projects like mine. There are so many classes passing through ```Isolates```.
I am really curious about this issue because everything was fine until flutter version 3.3.10.

**Comment by exaby73** [comment]
@dayoul Could you provide steps to reproduce this issue? I see you shared a repository in your original comment, but running it doesn't show any logs (Android) and in iOS I get `SendPort in ApplicationEngine is not set`

**Comment by dayoul** [comment]
@exaby73 Sorry my bad. Just fixed the code. Could you please try again on Android?

**Comment by exaby73** [comment]
So these are the logs printed in the console:

```console
I/flutter ( 5529): REQUEST from ActivityEngine::: 1
The Flutter DevTools debugger and profiler on sdk gphone64 arm64 is available at: http://127.0.0.1:9101?uri=http://127.0.0.1:50088/UXDdD92QEYw=/
I/flutter ( 5529): RESPONSE from ApplicationEngine::: 2
E/SurfaceSyncer( 5529): Failed to find sync for id=0
W/Parcel  ( 5529): Expecting binder but got null!
```

What am I looking at here? Sorry, but I haven't dug into the code :) But maybe you can explain better what the expected behavior is and what the actual is here

**Comment by dayoul** [comment]
@exaby73. Please check the Readme in the repository. 

## Expected Behavior
First load before hot restart
```
I/flutter (21335): REQUEST from ActivityEngine::: 1
I/flutter (21335): RESPONSE from ApplicationEngine::: 2
```
After hot restart
```
I/flutter (21335): REQUEST from ActivityEngine::: 1
I/flutter (21335): RESPONSE from ApplicationEngine::: 2
```
## Actual Behavior
First load before hot restart
```
I/flutter (21335): REQUEST from ActivityEngine::: 1
I/flutter (21335): RESPONSE from ApplicationEngine::: 2
```
After hot restart
```
I/flutter (24922): Invalid argument: is a regular instance: Instance of 'Data' #0      _SendPort._sendInternal (dart:isolate-patch/isolate_patch.dart:249:43)
I/flutter (24922): #1      _SendPort.send (dart:isolate-patch/isolate_patch.dart:230:5)
I/flutter (24922): #2      setApplicationEngineReceiver.<anonymous closure> (package:flutter_engine_group_project/isolate_executor.dart:23:27)
I/flutter (24922): #3      _rootRunUnary (dart:async/zone.dart:1414:13)
I/flutter (24922): #4      _CustomZone.runUnary (dart:async/zone.dart:1307:19)
I/flutter (24922): #5      _CustomZone.runUnaryGuarded (dart:async/zone.dart:1216:7)
I/flutter (24922): #6      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339:11)
I/flutter (24922): #7      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
I/flutter (24922): #8      _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:774:19)
I/flutter (24922): #9      _StreamController._add (dart:async/stream_controller.dart:648:7)
I/flutter (24922): #10     _StreamController.add (dart:async/stream_controller.dart:596:5)
I/flutter (24922): #11     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:192:26)
```

**Comment by exaby73** [comment]
## Triage report

With the code sample provided, I can reproduce this issue. Unfortunately, since I have limited understanding of `FlutterEngine` and `FlutterEngineGroup`, I am unable to investigate further and would love to get insights from the team here

Logs:

```console
I/flutter ( 5566): Invalid argument: is a regular instance: Instance of 'Data' #0      _SendPort._sendInternal (dart:isolate-patch/isolate_patch.dart:249:43)
I/flutter ( 5566): #1      _SendPort.send (dart:isolate-patch/isolate_patch.dart:230:5)
I/flutter ( 5566): #2      main.<anonymous closure> (package:flutter_engine_group_project/main.dart:25:35)
I/flutter ( 5566): <asynchronous suspension>
```

<details><summary><code>flutter doctor -v</code> (Master)</summary>

```
[!] Flutter (Channel master, 3.8.0-1.0.pre.4, on macOS 13.1 22C65 darwin-arm64, locale en-US)
    • Flutter version 3.8.0-1.0.pre.4 on channel master at /Users/nabeelparkar/fvm/versions/master
    ! Warning: `dart` on your path resolves to /opt/homebrew/Cellar/dart/2.18.6/libexec/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nabeelparkar/fvm/versions/master. Consider adding /Users/nabeelparkar/fvm/versions/master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision a752c2f154 (35 minutes ago), 2023-02-01 00:41:16 -0500
    • Engine revision 1703a39662
    • Dart version 3.0.0 (build 3.0.0-187.0.dev)
    • DevTools version 2.20.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nabeelparkar/Library/Android/sdk/
    • Platform android-33, build-tools 33.0.0
    • ANDROID_SDK_ROOT = /Users/nabeelparkar/Library/Android/sdk/
    • Java binary at: /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9123335/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /Applications/Brave Browser.app/Contents/MacOS/Brave Browser

[✓] Android Studio (version 2021.3)
    • Android Studio at /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9123335/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[!] Android Studio (version 2022.1)
    • Android Studio at /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/221.6008.13.2211.9477386/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[✓] VS Code (version 1.74.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.58.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.1 22C65 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Brave Browser 109.1.47.186

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 2 categories.
```

</details>

**Comment by dballance** [comment]
I have encountered the same issue. 

I establish a `FlutterEngineGroup` in my `Application` on Android. I do this so that I can start another isolate within a `Service` so that I can continue to run an isolate after the `MainActivity` is destroyed.

I then provide my own engine via `provideFlutterEngine` in my `MainActivity` by simply using `createAndRunDefaultEngine` from the `FlutterEngineGroup`. 

In my experience, both hot restarts and reloads no longer work. Hot reloads just "hang" and never reload, and hot restarts toss up the error listed in this issue.

My guess here is that the hot loading mechanisms aren't utilizing the same `FlutterEngineGroup` when reloading. 

Maybe flutter/engine#37822 will fix it? 

@gaaclarke - will the above resolve this on Android? If not - should there be a method to provide a `FlutterEngineGroup` similar to `provideFlutterEngine` so that hot reload mechanisms 

I could probably test this on the tip of flutter master. Unsure that the above PR landed in a stable release yet (doesn't appear to be by the state of docs on api.flutter.dev/javadoc).

**Comment by dballance** [comment]
I verified via the observatory some of the behavior here. 

When using an `FlutterEngineGroup`, each isolate spawned via a new engine is included in a single `IsolateGroup` on start. 

After a hot restart - the isolates are now running in separate `IsolateGroup` instances.

This seems like an oversight in the hot restart / hot reload mechanism - as the isolates are properly started in the same `IsolateGroup` and everything runs fine until a hot restart. Maybe it needs to call `reloadSources` on the groups, and not on the `Isolate` directly?

Reloads are still hanging - nothing to add on those.

**Comment by iazel** [comment]
Perhaps this issue is related to #147469 ?

Correct me if I'm wrong, but I believe that after hot-restarts isolates get killed, however the port mapping stays the same. If we consider that registering a port with the same name _will not_ overwrite it but rather keep the old one in place, the issue is now bound to happen.
